commands/utils.go

```diff
@@ -0,0 +1,209 @@
+	"github.com/ipfs/iptb/testbed/interfaces"
+)
+
+func isTerminatorPresent(c *cli.Context) bool {
+	argsParent := c.Parent().Args().Tail()
+	argsSelf := c.Args()
+
+	ls := len(argsSelf)
+	lp := len(argsParent)
+
+	term := lp - ls - 1
+
+	if lp > ls && argsParent[term] == "--" {
+		return true
+	}
+
+	return false
+}
+
+func parseAttrSlice(attrsraw []string) (map[string]interface{}, error) {
+	attrs := make(map[string]interface{})
+	for _, attr := range attrsraw {
+		parts := strings.Split(attr, ",")
+
+		if len(parts) == 1 {
+			attrs[parts[0]] = "true"
+		} else {
+			attrs[parts[0]] = strings.Join(parts[1:], ",")
+		}
+	}
+
+	return attrs, nil
+}
+
+func parseCommand(args []string, terminator bool) (string, []string) {
+	if terminator {
+		return "", args
+	}
+
+	if len(args) == 0 {
+		return "", []string{}
+	}
+
+	if len(args) == 1 {
+		return args[0], []string{}
+	}
+
+	arguments := args[1:]
+
+	if arguments[0] == "--" {
+		arguments = arguments[1:]
+	}
+
+	return args[0], arguments
+}
+
+func parseRange(s string) ([]int, error) {
+	if strings.HasPrefix(s, "[") && strings.HasSuffix(s, "]") {
+		ranges := strings.Split(s[1:len(s)-1], ",")
+		var out []int
+		for _, r := range ranges {
+			rng, err := expandDashRange(r)
+			if err != nil {
+				return nil, err
+			}
+
+			out = append(out, rng...)
+		}
+		return out, nil
+	}
+	i, err := strconv.Atoi(s)
+	if err != nil {
+		return nil, err
+	}
+
+	return []int{i}, nil
+}
+
+func expandDashRange(s string) ([]int, error) {
+	parts := strings.Split(s, "-")
+	if len(parts) == 0 {
+		i, err := strconv.Atoi(s)
+		if err != nil {
+			return nil, err
+		}
+		return []int{i}, nil
+	}
+	low, err := strconv.Atoi(parts[0])
+	if err != nil {
+		return nil, err
+	}
+
+	hi, err := strconv.Atoi(parts[1])
+	if err != nil {
+		return nil, err
+	}
+
+	var out []int
+	for i := low; i <= hi; i++ {
+		out = append(out, i)
+	}
+	return out, nil
+}
+
+func getNodes(desc string, tb testbed.BasicTestbed) ([]testbedi.Core, error) {
+	if len(desc) == 0 {
+		return tb.Nodes()
+	}
+
+	ndlist, err := parseRange(desc)
+	if err != nil {
+		return nil, err
+	}
+
+	var nodes []testbedi.Core
+	for _, n := range ndlist {
+		nd, err := tb.Node(n)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load node: %s", err)
+		}
+
+		nodes = append(nodes, nd)
+	}
+
+	return nodes, nil
+}
+
+type Result struct {
+	Node   int
+	Output testbedi.TBOutput
+	Error  error
+}
+
+type outputFunc func(testbedi.Core) (testbedi.TBOutput, error)
+
+func mapWithOutput(list []int, nodes []testbedi.Core, fn outputFunc) ([]Result, error) {
+	var wg sync.WaitGroup
+	var lk sync.Mutex
+	results := make([]Result, len(list))
+
+	// TODO(tperson) make sure list is in range
+	for i, n := range list {
+		wg.Add(1)
+		go func(i, n int, node testbedi.Core) {
+			defer wg.Done()
+			out, err := fn(node)
+
+			lk.Lock()
+			defer lk.Unlock()
+
+			results[i] = Result{
+				Node:   n,
+				Output: out,
+				Error:  errors.Wrapf(err, "node[%d]", n),
```

Comment:
will this this be incorrect if `list` is a range like [4-8]?

Reply:
yep!